driver-reactive-streams/src/main/com/mongodb/reactivestreams/client/internal/vault/ClientEncryptionImpl.java

```diff
@@ -183,6 +198,85 @@ public Publisher<RewrapManyDataKeyResult> rewrapManyDataKey(final Bson filter, f
+            Bson rawEncryptedFields = createCollectionOptions.getEncryptedFields();
+            if (rawEncryptedFields == null) {
+                throw new MongoConfigurationException(format("`encryptedFields` is not configured for the collection %s.", namespace));
+            }
+            CodecRegistry codecRegistry = options.getKeyVaultMongoClientSettings() == null
+                    ? MongoClientSettings.getDefaultCodecRegistry()
+                    : options.getKeyVaultMongoClientSettings().getCodecRegistry();
+            BsonDocument actualEncryptedFields = rawEncryptedFields.toBsonDocument(BsonDocument.class, codecRegistry).clone();
+            BsonValue fields = actualEncryptedFields.get("fields");
+            Mono<Void> createCollectionMono;
+            if (fields != null) {
+                if (!fields.isArray()) {
+                    throw new MongoConfigurationException(format("`encryptedFields` is incorrectly configured for the collection %s."
+                            + " `encryptedFields.fields` must be an array, but is %s type.", namespace, fields.getBsonType()));
+                }
+                String kmsProvider = createEncryptedCollectionParams.getKmsProvider();
+                DataKeyOptions dataKeyOptions = new DataKeyOptions();
+                BsonDocument masterKey = createEncryptedCollectionParams.getMasterKey();
+                if (masterKey != null) {
+                    dataKeyOptions.masterKey(masterKey);
+                }
+                String keyIdBsonKey = "keyId";
+                AtomicBoolean dataKeyMayBeCreated = new AtomicBoolean();
+                Iterable<Mono<SimpleImmutableEntry<BsonDocument, BsonBinary>>> publishersOfFieldsAndDataKeyIds = () -> fields.asArray()
+                        .stream()
+                        .filter(BsonValue::isDocument)
+                        .map(BsonValue::asDocument)
+                        .filter(field -> field.containsKey(keyIdBsonKey))
+                        .filter(field -> Objects.equals(field.get(keyIdBsonKey), BsonNull.VALUE))
+                        // here we rely on the `createDataKey` publisher being cold, i.e., doing nothing until it is subscribed to
+                        .map(field -> Mono.fromDirect(createDataKey(kmsProvider, dataKeyOptions))
+                                // This is the closest we can do with reactive streams to setting the `dataKeyMayBeCreated` flag
+                                // immediately before calling `createDataKey`.
+                                .doOnSubscribe(subscription -> dataKeyMayBeCreated.set(true))
+                                .map(dataKey -> new SimpleImmutableEntry<>(field, dataKey))
+                        )
+                        .iterator();
+                Flux<SimpleImmutableEntry<BsonDocument, BsonBinary>> publisherOfFieldsAndDataKeyIds = Flux.concat(publishersOfFieldsAndDataKeyIds)
```

Comment:
As a result of `Flux.concat` all data keys are created sequentially. This way we stop as soon as there is an error, instead possibly continuing to invoke `createDataKey` despite there already being an error.